### PR TITLE
Fail build when consumer and producer task run in parallel

### PR DESCRIPTION
**Pull request by wolfs**
In that case the build is really broken, so we can as well fail it
to expose the problem.

Follow up to #15411.

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Review by lptr — COMMENTED** [review]
Looks good, left a few comments.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 1]
```diff
@@ -124,9 +124,12 @@ private static boolean missesDependency(Node producer, Node consumer) {
         return true;
     }
 
-    private void collectValidationWarning(Node producer, Node consumer, TypeValidationContext validationContext) {
+    private void collectValidationWarningOrError(Node producer, Node consumer, TypeValidationContext validationContext) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java]: Renamed to `collectValidationProblem`.

File: subprojects/core/src/main/java/org/gradle/execution/plan/Node.java [hunk 2]
```diff
@@ -101,6 +101,10 @@ public boolean isExecuted() {
         return state == ExecutionState.EXECUTED;
     }
 
+    public boolean isExecuting() {
+        return state == ExecutionState.EXECUTING;
+    }
+
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/Node.java]: I changed the order.

File: subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java [hunk 3]
```diff
@@ -124,9 +124,12 @@ private static boolean missesDependency(Node producer, Node consumer) {
...
+        TypeValidationContext.Severity severity = producer.isExecuting() && consumer.isExecuting()
+            ? TypeValidationContext.Severity.ERROR
+            : TypeValidationContext.Severity.WARNING;
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java:132]: I added a comment.

**Review by lptr — APPROVED** [review]
LGTM.

**Pull request merged**